Sqoop init action (#268).

szewi opened the pull request.

karth295 commented on sqoop/README.MD:
```diff
@@ -0,0 +1,74 @@
+## Using this initialization action
+You can use this initialization action to create a new Dataproc cluster with Sqoop installed by:
+
+1. Uploading a copy of this initialization action (`sqoop.sh`) to [Google Cloud Storage](https://cloud.google.com/storage).
```
Sorry I haven't merged my README updates, but I'm planning to change the READMEs to just point at our GCS mirror: https://github.com/GoogleCloudPlatform/dataproc-initialization-actions/pull/262.

Please follow the style from that PR :)

szewi replied:
ok

karth295 commented on sqoop/README.MD:
````diff
@@ -0,0 +1,74 @@
+    ```bash
+    gcloud dataproc clusters create <CLUSTER_NAME> \
+        --initialization-action-timeout 15m --initialization-actions \
+        gs://dataproc-initialization-actions/cloud-sql-proxy/cloud-sql-proxy.sh,\
````
Nit: you don't need to `cloud-sql-proxy` init action here

pmkc replied:
Yes you do if you are using Cloud SQL as the RDBMS

szewi replied:
I am using sqoop to export data from cloud SQL and import data to BigTable in single sqoop job so I need it for that example.

karth295 commented on sqoop/README.MD:
```diff
@@ -0,0 +1,74 @@
+        gs://dataproc-initialization-actions/hive-hcatalog/hive-hcatalog.sh,\
+        gs://dataproc-initialization-actions/bigtable/bigtable.sh,\
+        gs://dataproc-initialization-actions/sqoop/sqoop.sh \
+        --single-node --master-machine-type n1-standard-4 --scopes sql-admin,\
```
Similarly omit `sql-admin` here

pmkc replied:
I always recommending using cloud-admin instead.

szewi replied:
Do you mean `cloud-platform` scope? 
cloud-admin results in an error: `Invalid service account scope: 'cloud-admin'`

karth295 commented on sqoop/sqoop.sh:
```diff
@@ -0,0 +1,73 @@
+
+readonly SQOOP_HOME='/usr/lib/sqoop'
+readonly SQOOP_CODE_LINK='https://github.com/apache/sqoop'
+readonly BIGTABLE_HBASE_CLIENT='bigtable-hbase-1.x-hadoop-1.3.0.jar'
```
If you're requiring the BT init action, is this still necessary?

szewi replied:
After adding HBase libs to classpath in bigtable init actions it is not needed now.

szewi replied:
Right now when sqoop is not compiled on the cluster, but downloaded as compiled jar from somewhere else, we need to provide a minimal set of dependencies, because during compilation several jars were downloaded and added to classpath [see: bin/configure-sqoop](https://github.com/szewi/sqoop/blob/sqoop-bigtable-client/bin/configure-sqoop#L206). There are several already available on dataproc clusters, but the only missing is avro-1.8.1, so I downloaded it manually and added to init action script.

szewi replied:
`bigtable-hbase-1.x-hadoop-1.3.0.jar` connector is a different story. When I use BigTable init action along with sqoop, the BigTable init action already installs this jar file in `/usr/lib/hbase/lib` and put those on the classpath. Then it's being passed to sqoop in [bin/configure-sqoop](https://github.com/szewi/sqoop/blob/sqoop-bigtable-client/bin/configure-sqoop#L175), but it doesn't work with class not found exception.(I even printed sqoop classpath and bigtable-hbase jar is there!) 
Solution is either to symlink or copy `bigtable-hbase-1.x-hadoop-1.3.0.jar` to /usr/lib/sqoop/lib directory, which is a natural way of installing sqoop connectors. 
I was thinking about automatically detecting which init actions were used to create a cluster and then if there was a `bigtable` init action getting/linknig right jar file, but this doesn't seem to me like a proper way of handling that, explicitily installing connector will present to users how to easily add more connectors, what do you think @karth295 ?

karth295 commented on sqoop/README.MD:
```diff
@@ -0,0 +1,74 @@
+# Sqoop 
+
+This initialization action installs [Sqoop](http://sqoop.apache.org/) on a [Google Cloud Dataproc](https://cloud.google.com/dataproc) cluster. Sqoop is being compiled from trunk branch of official [github](https://github.com/apache/sqoop) mirror.
```
Compiling from `trunk` is a little scary. We should at least compile from a release tag and just cherry-pick the necessary commit in.

@pmkc will we be able to backport [the commit](https://github.com/apache/sqoop/commit/e13dd21209c26316d43350a23f5d533321b61352) into Dataproc so we don't need to compile Sqoop in this init action?

pmkc replied:
This CL should not be compiling at all. It should certainly ignore BigTable until Sqoop is fixed in Bigtop with that patch

szewi replied:
Let me know when Sqoop will be available in Bigtop. I will test it and update this PR properly, so far I will just update docs.

pmkc commented on sqoop/sqoop.sh:
```diff
@@ -0,0 +1,73 @@
+function install_sqoop() {
+  git clone ${SQOOP_CODE_LINK} ${SQOOP_HOME} && cd ${SQOOP_HOME} && ant \
+    || err 'Compiling Sqoop from source code failed.'
+  echo "export PATH=\"${SQOOP_HOME}/bin:$PATH\"" >> /etc/profile
```
Source builds are pretty much forbidden in our init actions for performance reasons.

If Sqoop is not available as a binary distribution I would drop this PR and resubmit it when it's fixed in our Bigtop repo.

szewi replied:
@pmkc if you would like as temporary solution I can pack compiled sqoop in some lightweight docker container ? WDYT?

karth295 replied:
@szewi I think it will be a little while until this is in Bigtop. For now, I suggest we host a mirror on GCS.

I'd prefer not to build from trunk or any non-release commit, since non-release commits are not guaranteed to be stable. Please run `git cherry-pick e13dd21209c26316d43350a23f5d533321b61352` onto `release-1.4.7-rc0`, resolve the merge conflicts, run Sqoop's tests, and build a jar. I can then put that jar in the appropriate bucket.

If that's more complicated than just minor merge conflicts, then we can build directly from `e13dd21209c26316d43350a23f5d533321b61352` and host that.

Either way, this init action should be hermetic and should not build from trunk.

szewi replied:
@karth295 I cherry-picked `e13dd21209c26316d43350a23f5d533321b61352` onto release-1.4.7-rc0 resolved conflicts, but tests were not passing. I investigated this and decided to first cherry-pick commit `18445290810b1df035e06fb074064d6b9c1d6e90` which was solving the issues with tests and also allowed to apply then `e13dd21209c26316d43350a23f5d533321b61352` without conflicts. Then I compiled sqoop on a clean dataproc cluster and uploaded sqoop to publicly available bucket. Most probably you would like to move sqoop.jar to your own place on google storage, so please let me know about final location, then I will update URL properly. 

Sqoop is being delivered with `bin/` directory containing required executable scripts, so right now sqoop code is downloaded to /usr/lib/sqoop, but no compilation is done. I changed the source code location to my sqoop fork with the branch `sqoop-bigtable-client` containing changes I have applied, so the end user will have exact sources used for creating sqoop jar in /usr/lib/sqoop. Let me know if you are fine with this.

szewi commented:
@karth295 This is ready for another round of reviewing. Please take a look.

szewi commented:
@karth295 I have pushed docs fixes. 
@pmkc is the any update regarding adding scoop to Bigtop?

szewi commented:
Hi, @karth295 I have updated init action and README file, added comments and questions for decisions I made(not sure why not visible on main GitHub page, but just go to latest commit).
